libcalico-go/lib/backend/syncersv1/dedupebuffer/dedupe_buffer.go

```diff
@@ -0,0 +1,283 @@
+// deduplicate in-flight updates to the same keys.  It should only block for
+// short periods even if the downstream sink blocks for a long time.
+func (d *DedupeBuffer) OnUpdates(updates []api.Update) {
+	d.OnUpdatesKeysKnown(updates, nil)
+}
+
+// OnUpdatesKeysKnown is like OnUpdates, but it allows for the pre-serialised
+// keys of the KV pairs to be passed in.  If an entry in keys is "" or if keys
+// is shorter than updates the key will be computed.
+//
+// The updates and keys slices are not retained.
+func (d *DedupeBuffer) OnUpdatesKeysKnown(updates []api.Update, keys []string) {
+	debug := log.IsLevelEnabled(log.DebugLevel)
+	if debug {
+		log.WithField("numUpdates", len(updates)).Debug("Updates received")
+	}
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	if debug {
+		log.WithFields(log.Fields{
+			"queueLen": d.pendingUpdates.Len(),
+		}).Debug("Acquired lock")
+	}
+
+	queueWasEmpty := d.pendingUpdates.Len() == 0
+	for i, u := range updates {
+		var key string
+		if i < len(keys) {
+			// Have a cached key.
+			key = keys[i]
+		}
+		if key == "" {
+			// No key provided, calculate it.
+			var err error
+			key, err = model.KeyToDefaultPath(u.Key)
+			if err != nil {
+				// Shouldn't happen, we get out keys from Typha which has already
+				// Encoding them once!
+				log.WithError(err).WithField("key", u.Key).Error(
+					"Failed to generate default path for key.  Will skip this update.")
+				continue
+			}
+		}
+
+		if element, ok := d.keyToPendingUpdate[key]; ok {
+			// Already got an in-flight update for this key.
+			if u.Value == nil && !d.sentKeys.Contains(key) {
+				// This is a deletion, but the key in question never made it
+				// off the queue, remove it entirely.
+				if debug {
+					log.WithField("key", key).Debug("Key deleted before being sent.")
+				}
+				delete(d.keyToPendingUpdate, key)
+				d.pendingUpdates.Remove(element)
+			} else {
+				// Update to a key that's already on the queue, swap in the
+				// most recent value.
+				if debug {
+					log.WithField("key", key).Debug("Key updated before being sent.")
+				}
+				usk := element.Value.(updateWithStringKey)
+				usk.update = u
+				element.Value = usk
+			}
+		} else {
+			// No in-flight entry for this key.  Add to queue and record that
+			// it's in flight.
+			if debug {
+				log.WithField("key", key).Debug("No in flight value for key, adding to queue.")
+			}
+			element = d.pendingUpdates.PushBack(updateWithStringKey{
+				key:    key,
+				update: u,
+			})
+			d.keyToPendingUpdate[key] = element
+		}
+	}
+	queueNowEmpty := d.pendingUpdates.Len() == 0
+	if queueWasEmpty && !queueNowEmpty {
+		// Only need to signal when the first item goes on the queue.
+		if debug {
+			log.Debug("Queue transitioned to non-empty; signalling.")
+		}
+		d.cond.Signal()
+	}
+}
+
+func (d *DedupeBuffer) SendToSinkForever(sink api.SyncerCallbacks) {
+	for !d.stopped {
+		d.sendNextBatchToSink(sink)
+	}
```

Comment:
Is there subtlety hiding here in the fact that the code for sendNextBatchToSink is not inlined?

- Is it OK to query `d.stopped` outside the lock?
- Does it matter that we drop and reacquire the lock between iterations of the above loop?  Is that needed to allow upstream to add stuff to our queue, or would the dropping in dropLockAndSendBatch already suffice for that?

Depending on the answers, I'd slightly prefer inlining here, with one overall
```
d.lock.Lock()
defer d.lock.Unlock()
```
at the start of this function.